src/python/grpcio/grpc/_cython/_cygrpc/aio/server.pyx.pxi

```diff
@@ -781,6 +781,39 @@ cdef CallbackFailureHandler SERVER_SHUTDOWN_FAILURE_HANDLER = CallbackFailureHan
     InternalError)
 
 
+cdef class _ConcurrentRpcLimiter:
+
+    def __cinit__(self, int maximum_concurrent_rpcs, object loop):
+        if maximum_concurrent_rpcs <= 0:
+            raise ValueError("maximum_concurrent_rpcs should be a postive integer")
+        self._maximum_concurrent_rpcs = maximum_concurrent_rpcs
+        self._active_rpcs = 0
+        self._active_rpcs_condition = asyncio.Condition()
+        self._loop = loop
+
+    async def check_before_request_call(self):
+        await self._active_rpcs_condition.acquire()
+        try:
+            predicate = lambda: self._active_rpcs < self._maximum_concurrent_rpcs
+            await self._active_rpcs_condition.wait_for(predicate)
+            self._active_rpcs += 1
+        finally:
+            self._active_rpcs_condition.release()
+
+    def decrease_once_finished(self, object rpc_task):
+        def decrease_active_rpcs_count(_):
+            async def decrease_active_rpcs_count_with_lock():
+                await self._active_rpcs_condition.acquire()
+                try:
+                    self._active_rpcs -= 1
+                    self._active_rpcs_condition.notify()
```

Comment:
I was wondering about coroutine starvation here.There doesn't seem to be an API guarantee on the order in which waiting coroutines are woken up by a call to `notify`. If any arbitrary ordering is valid, then it's possible for a waiting coroutine to never make progress. Based on the [`asyncio.Condition` implementation](https://github.com/python/cpython/blob/79d2e62c008446fbbc6f264bb8a30e2d38b6ff58/Lib/asyncio/locks.py#L337), it looks like this shouldn't happen in practice.

Reply:
Yeah, we don't really have memory order here. Feel free to suggest improvements.

Reply:
No, I think what you have is sufficient. I just wanted to point this out in case cpython changes and we see a problem stemming from this in the future.